addons/web/static/src/core/assets.js

```diff
@@ -65,118 +67,196 @@ export const loadCSS = memoize(function loadCSS(url) {
+ *      The list of assets to load. Each list item may be a string (the xmlID
+ *      of the asset to load) or a list of strings. The first level is loaded
+ *      sequentially (so use this if the order matters) while the assets in
+ *      inner lists are loaded in parallel (use this for efficiency but only
+ *      if the order does not matter, should rarely be the case for assets).
+ * @param {string[]} [desc.cssLibs=[]]
+ *      The list of CSS files to load. They will all be loaded in parallel but
+ *      put in the DOM in the given order (only the order in the DOM is used
+ *      to determine priority of CSS rules, not loaded time).
+ * @param {Array<string|string[]>} [desc.jsLibs=[]]
+ *      The list of JS files to load. Each list item may be a string (the URL
+ *      of the file to load) or a list of strings. The first level is loaded
+ *      sequentially (so use this if the order matters) while the files in inner
+ *      lists are loaded in parallel (use this for efficiency but only
+ *      if the order does not matter).
+ * @param {string[]} [desc.cssContents=[]]
+ *      List of inline styles to add after loading the CSS files.
+ * @param {string[]} [desc.jsContents=[]]
+ *      List of inline scripts to add after loading the JS files.
  *
- * @deprecated
- * @param {string} xmlid The xmlid of the template that defines the public asset
- * @param {ORM} orm An ORM object capable of calling methods on models
- * @returns {Promise<void>} Resolved when the contents of the asset is loaded
+ * @returns {Promise}
  */
-export const loadPublicAsset = memoize(async function loadPublicAsset(xmlid, orm) {
-    const xml = await orm.call("ir.ui.view", "render_public_asset", [xmlid]);
-    const doc = new DOMParser().parseFromString(`<xml>${xml}</xml>`, "text/xml");
-    return Promise.all([
-        ...[...doc.querySelectorAll("link[href]")].map((el) => loadCSS(el.getAttribute("href"))),
-        ...[...doc.querySelectorAll("script[src]")].map((el) => loadJS(el.getAttribute("src"))),
-    ]);
+assets.loadBundle = memoize(async function loadBundle(desc) {
+    // Load css in parallel
+    const promiseCSS = Promise.all((desc.cssLibs || []).map(assets.loadCSS)).then(function () {
+        if (desc.cssContents && desc.cssContents.length) {
+            const style = document.createElement("style");
+            style.innerHTML = desc.cssContents.join('\n');
+            document.head.appendChild(style);
+        }
+    });
+    // Load JavaScript (don't wait the css loading)
+    for (const urlData of desc.jsLibs || []) {
+        if (typeof urlData === 'string') {
+            // serial loading
+            await assets.loadJS(urlData);
+            // Wait template if the JavaScript come from bundle.
+            const bundle = urlData.match(/\/web\/assets\/.*\/([^\/]+?)(\.min)?\.js/);
+            if (bundle) {
+                await odoo.ready(bundle[1] + '.bundle.xml');
+            }
```

Comment:
Is this really necessary? Can't we flatten all the JS sources extracting XML information, load all the JS, then load all the XML?

Reply:
Yes because the module is put at the end of the file, the templates are loaded at an unknown time. We want to know when they are loaded before mounting elements.